performance_manager/lib/l2_headways.py

```diff
@@ -0,0 +1,383 @@
+    vp_timestamp_query = sa.select(
+        sa.func.min(VehiclePositionEvents.timestamp_start).label(
+            "timestamp_start"
+        ),
+    ).join(
+        FullTripEvents,
+        FullTripEvents.fk_vp_stopped_event == VehiclePositionEvents.pk_id,
+    )
+
+    # get maximum updated_on timestamp from headways table
+    max_headway_update_query = sa.select(sa.func.max(Headways.updated_on))
+    # mypy error: error: "BaseCursorResult" has no attribute "fetchone"
+    max_update_dt = db_manager.execute(max_headway_update_query).fetchone()[0]  # type: ignore
+
+    # will only be None if no records exist in Headways table
+    # if records do exist, add constraing to vehicle position and trip update
+    # timestamp_start queries
+    if max_update_dt is not None:
+        tu_timestamp_query = tu_timestamp_query.where(
+            FullTripEvents.updated_on > max_update_dt
+        )
+        vp_timestamp_query = vp_timestamp_query.where(
+            FullTripEvents.updated_on > max_update_dt
+        )
+
+    # union query so all results are combined
+    min_timestamp_start_cte = tu_timestamp_query.union(vp_timestamp_query).cte(
+        name="min_timestamp"
+    )
+
+    # get minimum timestamp_start value for records needing to be added to
+    # headways table
+    min_timestamp_start_query = sa.select(
+        sa.func.min(min_timestamp_start_cte.c.timestamp_start)
+    )
+
+    # this will return a valid timestamp_start integer or None, if no records
+    # need to be added to headways table
+    # mypy error: error: "BaseCursorResult" has no attribute "fetchone"
+    min_timestamp_start = db_manager.execute(min_timestamp_start_query).fetchone()[0]  # type: ignore
+
+    # pull overlapping timestamp start values going back 6 Hours
+    # this effecitvely sets a maxium headway value of 6 hours
+    # it is assumed that headways over 6 hours are not really headways and do not
+    # need to be recorded
+    if isinstance(min_timestamp_start, int):
+        min_timestamp_start -= 60 * 60 * 6
+
+    # return None or int value
+    return min_timestamp_start
+
+
+def load_temp_headways(db_manager: DatabaseManager) -> int:
+    """
+    perform update on headways table
+
+    """
+    process_logger = ProcessLogger("l2_load_temp_headways")
+    process_logger.log_start()
+
+    # get minimum timestamp_start value for records that need to be processed
+    # for headways table
+    # will return None if no headways need to be processed
+    min_timestamp_start = get_min_timestamp(db_manager)
+    if min_timestamp_start is None:
+        process_logger.add_metadata(temp_rows_inserted=0)
+        process_logger.log_complete()
+        return 0
+
+    # headways will be aggregated by the following fields:
+    # - direction_id
+    # - route_id
+    # - start_date
+    # - parent_station
+    #
+    # query for stop event records from vehcile positions that require headways
+    # processing
+    vp_query = (
```

Comment:
why are these queries rather than CTEs?

Reply:
CTE's are used when you'll reference the result of a subquery later in the your SQL statement. 

In this instance, these two queries are joined with a `UNION ALL`. The joined results are used as a CTE because they are used later in the query.